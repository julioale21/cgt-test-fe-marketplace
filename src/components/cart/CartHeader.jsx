import React from 'react';
import PropTypes from 'prop-types';
import { Box, Typography, CardHeader } from '@mui/material';
import { OutlinedButton } from '../buttons/OutlinedButton';


const CartHeader = ({ itemCount, onClearCart }) => (
  <CardHeader
    title={
      <Box sx={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center' }}>
        <Typography variant="h6">Shopping Cart ({itemCount} items)</Typography>
        <OutlinedButton size="small" onClick={onClearCart}>
          Clear Cart
        </OutlinedButton>
      </Box>
    }
  />
);

CartHeader.propTypes = {
  itemCount: PropTypes.number.isRequired,
  onClearCart: PropTypes.func.isRequired
};

export { CartHeader };
