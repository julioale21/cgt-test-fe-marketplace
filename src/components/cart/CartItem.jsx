import React from 'react';
import PropTypes from 'prop-types';
import { Box, Typography, IconButton } from '@mui/material';
import { Add as AddIcon, Remove as RemoveIcon, Delete as DeleteIcon } from '@mui/icons-material';

const CartItem = ({ item, onIncrement, onDecrement, onRemove }) => (
  <Box
    sx={{
      py: 3,
      display: 'flex',
      gap: 3
    }}
  >
    <Box
      component="img"
      src={item.image}
      alt={item.name}
      sx={{
        width: '80px',
        height: '80px',
        objectFit: 'cover',
        borderRadius: 2
      }}
    />

    <Box
      sx={{
        flexGrow: 1,
        display: 'flex',
        flexDirection: 'column',
        justifyContent: 'space-between'
      }}
    >
      <Box
        sx={{
          display: 'flex',
          justifyContent: 'space-between',
          alignItems: 'flex-start'
        }}
      >
        <Box>
          <Typography variant="h6" sx={{ fontWeight: 500 }}>
            {item.name}
          </Typography>
          <Typography variant="h5" sx={{ mt: 1, fontWeight: 500 }}>
            ${item.price.toFixed(2)}
          </Typography>
        </Box>

        <Box
          sx={{
            display: 'flex',
            alignItems: 'center',
            gap: 1
          }}
        >
          <IconButton
            onClick={onDecrement}
            size="small"
            sx={{
              color: 'action.active',
              '&:hover': {
                color: 'primary.main',
                bgcolor: 'primary.lighter'
              }
            }}
          >
            <RemoveIcon />
          </IconButton>
          <Typography
            sx={{
              minWidth: '32px',
              textAlign: 'center'
            }}
          >
            {item.quantity}
          </Typography>
          <IconButton
            onClick={onIncrement}
            size="small"
            sx={{
              color: 'action.active',
              '&:hover': {
                color: 'primary.main',
                bgcolor: 'primary.lighter'
              }
            }}
          >
            <AddIcon />
          </IconButton>
          <IconButton
            onClick={onRemove}
            size="small"
            color="error"
            sx={{
              ml: 2,
              '&:hover': {
                bgcolor: 'error.lighter'
              }
            }}
          >
            <DeleteIcon />
          </IconButton>
        </Box>
      </Box>
    </Box>
  </Box>
);

CartItem.propTypes = {
  item: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    image: PropTypes.string.isRequired,
    quantity: PropTypes.number.isRequired
  }).isRequired,
  onIncrement: PropTypes.func.isRequired,
  onDecrement: PropTypes.func.isRequired,
  onRemove: PropTypes.func.isRequired
};

export { CartItem };
