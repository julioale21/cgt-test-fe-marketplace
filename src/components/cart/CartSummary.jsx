import React from 'react';
import { Box, Card, CardContent, CardHeader, Typography, Divider, Stack } from '@mui/material';
import PropTypes from 'prop-types';

import {
  ArrowBack as ArrowBackIcon,
  CreditCard as CreditCardIcon,
  LocalShipping as ShippingIcon
} from '@mui/icons-material';

import {
  CartSummaryRow,
  ContainedButton,
  OutlinedButton,
  PaymentMethodIcon
} from '../../components';

const CartSummary = ({
  subtotal,
  shipping = 0,
  taxes,
  finalTotal,
  onCheckout,
  onContinueShopping
}) => {
  const formatAmount = (amount) => {
    const nonNegativeAmount = Math.max(0, amount);
    return `$${nonNegativeAmount.toLocaleString('en-US', {
      minimumFractionDigits: 2,
      maximumFractionDigits: 2
    })}`;
  };

  return (
    <Card sx={{ position: { md: 'sticky' }, top: { md: '24px' } }}>
      <CardHeader title="Order Summary" />
      <Divider />
      <CardContent>
        <Stack spacing={2}>
          <CartSummaryRow label="Subtotal" value={formatAmount(subtotal)} />
          <CartSummaryRow
            label="Shipping"
            value={shipping === 0 ? 'Free' : formatAmount(shipping)}
          />
          <CartSummaryRow label="Estimated tax" value={formatAmount(taxes)} />
          <Divider />
          <CartSummaryRow
            label="Total"
            value={formatAmount(finalTotal)}
            variant="h6"
            labelColor="text.primary"
          />

          <ContainedButton
            fullWidth
            onClick={onCheckout}
            startIcon={<CreditCardIcon />}
            sx={{ py: 1.5 }}
          >
            Continue to Checkout
          </ContainedButton>

          <OutlinedButton
            fullWidth
            onClick={onContinueShopping}
            startIcon={<ArrowBackIcon />}
            sx={{ py: 1.5 }}
          >
            Continue Shopping
          </OutlinedButton>

          <Box sx={{ mt: 2 }}>
            <Typography variant="subtitle2" gutterBottom sx={{ mb: 1 }}>
              Secure Checkout
            </Typography>
            <Stack
              direction="row"
              spacing={2}
              sx={{
                py: 2,
                px: 1,
                bgcolor: 'grey.50',
                borderRadius: 1,
                justifyContent: 'center'
              }}
            >
              <PaymentMethodIcon
                src="https://cdn.visa.com/v2/assets/images/logos/visa/blue/logo.png"
                alt="Visa"
              />
              <PaymentMethodIcon
                src="https://cdn.worldvectorlogo.com/logos/mastercard-7.svg"
                alt="Mastercard"
              />
              <PaymentMethodIcon
                src="https://www.paypalobjects.com/digitalassets/c/website/logo/full-text/pp_fc_hl.svg"
                alt="PayPal"
              />
            </Stack>
          </Box>

          <Box sx={{ mt: 2 }}>
            <Typography
              variant="body2"
              color="text.secondary"
              sx={{ display: 'flex', alignItems: 'center', gap: 1 }}
            >
              <ShippingIcon fontSize="small" />
              Free shipping on all orders
            </Typography>
          </Box>
        </Stack>
      </CardContent>
    </Card>
  );
};

CartSummary.propTypes = {
  subtotal: PropTypes.number.isRequired,
  shipping: PropTypes.number,
  taxes: PropTypes.number.isRequired,
  finalTotal: PropTypes.number.isRequired,
  onCheckout: PropTypes.func.isRequired,
  onContinueShopping: PropTypes.func.isRequired
};

export { CartSummary };
