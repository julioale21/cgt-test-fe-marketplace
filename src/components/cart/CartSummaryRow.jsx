import React from 'react';
import PropTypes from 'prop-types';
import { Box, Typography } from '@mui/material';

const CartSummaryRow = ({
  label,
  value,
  variant = 'body1',
  labelColor = 'text.secondary',
  valueColor = 'text.primary'
}) => {
  return (
    <Box sx={{ display: 'flex', justifyContent: 'space-between' }}>
      <Typography color={labelColor} variant={variant}>
        {label}
      </Typography>
      <Typography color={valueColor} variant={variant}>
        {value}
      </Typography>
    </Box>
  );
};

CartSummaryRow.propTypes = {
  label: PropTypes.string.isRequired,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  variant: PropTypes.oneOf([
    'body1',
    'body2',
    'h1',
    'h2',
    'h3',
    'h4',
    'h5',
    'h6',
    'subtitle1',
    'subtitle2',
    'caption',
    'overline'
  ]),
  labelColor: PropTypes.string,
  valueColor: PropTypes.string
};

export { CartSummaryRow };
