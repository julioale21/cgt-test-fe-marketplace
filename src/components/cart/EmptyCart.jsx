import React from 'react';
import PropTypes from 'prop-types';
import { Card, CardContent, Typography } from '@mui/material';
import { ShoppingCart as ShoppingCartIcon, ArrowBack as ArrowBackIcon } from '@mui/icons-material';
import { OutlinedButton } from '../buttons/OutlinedButton';

const EmptyCart = ({ onContinueShopping }) => (
  <Card sx={{ maxWidth: 800, mx: 'auto', mt: 4 }}>
    <CardContent sx={{ textAlign: 'center', py: 4 }}>
      <ShoppingCartIcon sx={{ fontSize: 48, color: 'text.secondary', mb: 2 }} />
      <Typography color="text.secondary" gutterBottom>
        Your cart is empty
      </Typography>
      <OutlinedButton startIcon={<ArrowBackIcon />} onClick={onContinueShopping} sx={{ mt: 2 }}>
        Start Shopping
      </OutlinedButton>
    </CardContent>
  </Card>
);

EmptyCart.propTypes = {
  onContinueShopping: PropTypes.func.isRequired
};

export { EmptyCart };
