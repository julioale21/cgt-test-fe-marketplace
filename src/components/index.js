export { ContainedButton } from "./buttons/ContainedButton";
export { Header } from "./shared/Header";
export { Layout } from "./Layout";
export { OutlinedButton } from "./buttons/OutlinedButton";
export { ProductList } from "./ProductList";
export {ProductCard} from "./ProductCard";
