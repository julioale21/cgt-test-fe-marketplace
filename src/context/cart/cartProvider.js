import { useReducer, useCallback } from 'react';
import { cartReducer } from './cartReducer';
import { initialState, CART_ACTIONS } from './types';
import { CartContext } from './cartContext';

export const CartProvider = ({ children }) => {
  const [state, dispatch] = useReducer(cartReducer, initialState);

  const addToCart = useCallback(({ product, quantity }) => {
    dispatch({ 
      type: CART_ACTIONS.ADD_TO_CART, 
      payload: { product, quantity }
    });
  }, []);

  const removeFromCart = useCallback((productId) => {
    dispatch({ 
      type: CART_ACTIONS.REMOVE_FROM_CART, 
      payload: productId 
    });
  }, []);

  const updateQuantity = useCallback((productId, quantity) => {
    dispatch({
      type: CART_ACTIONS.UPDATE_QUANTITY,
      payload: { id: productId, quantity }
    });
  }, []);

  const clearCart = useCallback(() => {
    dispatch({ type: CART_ACTIONS.CLEAR_CART });
  }, []);

  const value = {
    items: state.items,
    total: state.total,
    addToCart,
    removeFromCart,
    updateQuantity,
    clearCart
  };

  return (
    <CartContext.Provider value={value}>
      {children}
    </CartContext.Provider>
  );
};