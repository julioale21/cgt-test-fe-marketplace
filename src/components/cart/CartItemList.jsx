import React from 'react';
import PropTypes from 'prop-types';
import { Stack } from '@mui/material';
import { CartItem } from './CartItem';

const CartItemList = ({ items, onIncrement, onDecrement, onRemove }) => (
  <Stack spacing={2}>
    {items.map((item) => (
      <CartItem
        key={item.id}
        item={item}
        onIncrement={() => onIncrement(item)}
        onDecrement={() => onDecrement(item)}
        onRemove={() => onRemove(item.id)}
      />
    ))}
  </Stack>
);

CartItemList.propTypes = {
  items: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      price: PropTypes.number.isRequired,
      image: PropTypes.string.isRequired,
      quantity: PropTypes.number.isRequired
    })
  ).isRequired,
  onIncrement: PropTypes.func.isRequired,
  onDecrement: PropTypes.func.isRequired,
  onRemove: PropTypes.func.isRequired
};

export { CartItemList };
