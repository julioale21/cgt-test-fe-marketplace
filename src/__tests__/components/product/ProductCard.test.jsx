import React from 'react';
import { render, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { ProductCard } from '../../../components';
import { CartProvider } from '../../../context/cart/cartProvider';

const mockProduct = {
  id: 1,
  name: '3D Modern Chair',
  price: 149.99,
  description: 'High-quality 3D model of a modern minimalist chair',
  image: '/path/to/image.jpg',
  category: 'Furniture',
  rating: 4.5,
  format: ['FBX', 'OBJ', 'MAX'],
  polygons: '24k',
  textures: '4K PBR'
};

const renderWithProviders = (component) => {
  return render(
    <CartProvider>
      <BrowserRouter>{component}</BrowserRouter>
    </CartProvider>
  );
};

describe('ProductCard', () => {
  it('should match snapshot', () => {
    const props = { product: mockProduct };

    const { container } = renderWithProviders(<ProductCard {...props} />);

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot when no product is provided', () => {
    const props = { product: null };

    const { container } = renderWithProviders(<ProductCard {...props} />);

    expect(container).toMatchSnapshot();
  });

  it('should render all product information correctly', () => {
    const props = { product: mockProduct };

    renderWithProviders(<ProductCard {...props} />);

    expect(screen.getByText(mockProduct.name)).toBeInTheDocument();
    expect(screen.getByText(mockProduct.description)).toBeInTheDocument();
    expect(screen.getByText(`$${mockProduct.price.toFixed(2)}`)).toBeInTheDocument();
    expect(screen.getByText(`Rating: ${mockProduct.rating}/5`)).toBeInTheDocument();
    expect(screen.getByText(`Polygons: ${mockProduct.polygons}`)).toBeInTheDocument();
    expect(screen.getByText(`Textures: ${mockProduct.textures}`)).toBeInTheDocument();
    expect(screen.getByText(`Formats: ${mockProduct.format.join(', ')}`)).toBeInTheDocument();
  });

  it('should render product image with correct attributes', () => {
    const props = { product: mockProduct };

    renderWithProviders(<ProductCard {...props} />);

    const image = screen.getByRole('img');
    expect(image).toHaveAttribute('src', mockProduct.image);
    expect(image).toHaveAttribute('alt', mockProduct.name);
  });

  it('should render action buttons', () => {
    const props = { product: mockProduct };

    renderWithProviders(<ProductCard {...props} />);

    expect(screen.getByText('View Details')).toBeInTheDocument();
    expect(screen.getByText('Add to Cart')).toBeInTheDocument();
  });

  it('should have correct link to product details', () => {
    const props = { product: mockProduct };

    renderWithProviders(<ProductCard {...props} />);

    const detailsLink = screen.getByText('View Details').closest('a');
    expect(detailsLink).toHaveAttribute('href', `/products/${mockProduct.id}`);
  });

  it('should return null when no product is provided', () => {
    const props = { product: null };

    const { container } = renderWithProviders(<ProductCard {...props} />);

    expect(container).toBeEmptyDOMElement();
  });
});
