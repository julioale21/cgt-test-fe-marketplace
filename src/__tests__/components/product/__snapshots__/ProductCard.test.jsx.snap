// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard should match snapshot 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-3 css-13i4rnv-MuiGrid-root"
  >
    <div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-lu81qf-MuiPaper-root-MuiCard-root"
        style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
      >
        <div>
          <img
            alt="3D Modern Chair"
            class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-hh48uz-MuiCardMedia-root"
            height="200"
            src="/path/to/image.jpg"
          />
        </div>
        <div
          class="MuiCardContent-root css-x4lmbv-MuiCardContent-root"
        >
          <div
            style="opacity: 0; transform: translateY(20px);"
          >
            <h2
              class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-1wmifpn-MuiTypography-root"
            >
              3D Modern Chair
            </h2>
            <p
              class="MuiTypography-root MuiTypography-body2 css-19b5ozw-MuiTypography-root"
            >
              High-quality 3D model of a modern minimalist chair
            </p>
            <h6
              class="MuiTypography-root MuiTypography-h6 css-11ary1m-MuiTypography-root"
            >
              $
              149.99
            </h6>
            <p
              class="MuiTypography-root MuiTypography-body2 css-1wle3ir-MuiTypography-root"
            >
              Rating: 
              4.5
              /5
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2 css-1wle3ir-MuiTypography-root"
            >
              Polygons: 
              24k
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2 css-1wle3ir-MuiTypography-root"
            >
              Textures: 
              4K PBR
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2 css-tmy7vb-MuiTypography-root"
            >
              Formats: 
              FBX, OBJ, MAX
            </p>
          </div>
        </div>
        <div
          class="MuiCardActions-root MuiCardActions-spacing css-gfsvm0-MuiCardActions-root"
        >
          <div
            tabindex="0"
          >
            <a
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-1tzdgtk-MuiButtonBase-root-MuiButton-root"
              href="/products/1"
              id=":r0:"
              tabindex="0"
            >
              View Details
            </a>
          </div>
          <div
            tabindex="0"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-uwf23-MuiButtonBase-root-MuiButton-root"
              id=":r1:"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeMedium css-1sh91j5-MuiButton-startIcon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1wdc28j-MuiSvgIcon-root"
                  data-testid="ShoppingCartIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2M1 2v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.14 0-.25-.11-.25-.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.58-6.49c.08-.14.12-.31.12-.48 0-.55-.45-1-1-1H5.21l-.94-2zm16 16c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2"
                  />
                </svg>
              </span>
              Add to Cart
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ProductCard should match snapshot when no product is provided 1`] = `<div />`;
